### PR TITLE
TM26 test balancing refactor

### PR DESCRIPTION
**Pull request by r3domfox**
Work in progress, raising it early so interested parties can take a look at it.

Moved some more groovy pieces to Java, refactored some of the existing Java pieces a bit.

Needs further organisation, Javadocs etc. - will update this description once it's all done.

### CODE REVIEW

File: buildSrc/src/main/java/net/corda/testing/balancing/BucketingAllocator.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import java.util.*;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+/**
+ * Accumulates test sources, and uses provided test timings to generate a test plan once all sources are captured.
+ */
+class BucketingAllocator {
+
+    static @NotNull BucketingAllocator create(int forkCount, @NotNull Supplier<Stream<TestTiming>> timingsProvider) {
```
Comment (barrylapthorn) [inline buildSrc/src/main/java/net/corda/testing/balancing/BucketingAllocator.java:16]: My version is just `Supplier<Tests>`, where `Tests` encapsulates the downloaded historic test durations from artifactory.

File: buildSrc/src/main/java/net/corda/testing/KubesReporting.java [hunk 2]
```diff
@@ -133,62 +146,77 @@ public void setTestResultDirs(Iterable<File> testResultDirs) {
...
+            containersWithNonZeroReturnCodes.forEach(this::reportContainerOutput);
+        }
+
+        String reportUrl = new ConsoleRenderer().asClickableFileUrl(new File(destinationDir, "index.html"));
+        throw new GradleException("remote build failed, check test report at " + reportUrl);
+    }
+
+    private void reportContainerOutput(KubePodResult podResult) {
+        try {
+            System.out.println("\n##### CONTAINER OUTPUT START #####");
```
Comment (barrylapthorn) [inline buildSrc/src/main/java/net/corda/testing/KubesReporting.java:194]: Replace the `println`?  IIRC you did that somewhere else in the PR.

File: buildSrc/src/main/java/net/corda/testing/balancing/BucketingAllocatorTask.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+import java.util.Objects;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class BucketingAllocatorTask extends DefaultTask {
+
+    private static final String TEST_NAME = "Test Name";
+    private static final String DURATION = "Duration(ms)";
+    private static final String DEFAULT_TESTING_TEST_TIMES_CSV = "testing/test-times.csv";
+
```
Comment (barrylapthorn) [inline buildSrc/src/main/java/net/corda/testing/balancing/BucketingAllocatorTask.java:27]: All of the 'load from csv' code is replaced with get-from-artifactory in my PR.

**Comment by rnicoll** [comment]
Dries has a copy for reference but this no longer is on a path to be merged as-is, closing.

**Pull request closed without merging**